### ISSUE TITLE
In the benchmark studies, how are the draft tokens generated?

### ISSUE THREAD

**Issue by jivanph** [alipay/PainlessInferenceAcceleration#9]
I read with great interest your paper 'Lookahead: An Inference Acceleration Framework for Large Language Model with Lossless Generation Accuracy'.

In essence, the paper proposes a tree data structure to verify proposed draft tokens, and in this way speed up inference.

Unfortunately, it's not clear to me from the paper how these draft tokens were generated when establishing benchmark results for LookAhead-Parallel and LookAhead-Hierarchical.

I understand the focus on the paper is on how to handle a set of draft tokens (perhaps as a single branch, perhaps in parallel, or perhaps in a hierarchical manner). But the origin of the draft tokens in the benchmark results remains unclear to me.

**Comment by jivanph** [comment]
A related  question in regards to the benchmark studies, what was the sampling mechanism used to accept tokens? Was it greedy sampling?

**Comment by zheyishine** [comment]
To Q1: The draft tokens are generated from a cached trie tree (each node is a token id). Currently the trie tree is constructed from prompts and responses on-the-fly,  therefore it is friendly for deployment (neither additonal assist models nor head training), and works pretty well in real-life scenarios. A bit more, we also have probed Jaccobi iteraction to  yield addtional drafts and will integrate into the repo soon(even though its speedup is marginal).
To Q2: Yes, we use the greedy strategy in our benchmarks.

**Comment by jivanph** [comment]
Thank you for your responses. I understand the tokens are passed as a tree object. But my question is, how did you choose which tokens to use in the benchmark? What do you mean by "responses on-the-fly"?

**Comment by zheyishine** [comment]
We choose tokens not only from responses, but also from prompts.

**Comment by jivanph** [comment]
Could you explain a little bit further how you choose the tokens?

**Comment by zheyishine** [comment]
In the benchmark, we first generate responses for samples from dev set, and put the responses into a global trie tree, then we evaluate each prompt in the test set( all the samples are different from the ones from the dev set). For each query in the test set, we first put the query into a global trie tree , and the generate tokens are also put into the trie tree on-the-fly.

**Comment by jivanph** [comment]
Thank you for your response. Could you please point me which part of the code is in charge of the verification step?

**Comment by zheyishine** [comment]
Lines from [here](https://github.com/alipay/PainlessInferenceAcceleration/blob/6280cb2f097ba0bc6bc423ab910b9de7ddbe3bf2/pia/lookahead/common/pretrained_model.py#L807) to [here](https://github.com/alipay/PainlessInferenceAcceleration/blob/6280cb2f097ba0bc6bc423ab910b9de7ddbe3bf2/pia/lookahead/common/pretrained_model.py#L861) are used for vefification of tree drafts.

Our `lookahead` decoding can not generate exactly the same response as the generation mode `SAMPLE` in transformers, due to random sampling( i.e., caused by `torch.multinomial`). We guarantee the same distribution by following the decoding steps of generation mode`SAMPLE`. Our implement is aligned with the  generation mode `ASSISTED_GENERATION` in `transformers`. 
Details can be found with lines from [here](https://github.com/alipay/PainlessInferenceAcceleration/blob/6280cb2f097ba0bc6bc423ab910b9de7ddbe3bf2/pia/lookahead/common/pretrained_model.py#L835) to [here](https://github.com/alipay/PainlessInferenceAcceleration/blob/6280cb2f097ba0bc6bc423ab910b9de7ddbe3bf2/pia/lookahead/common/pretrained_model.py#L840).

**Comment by fuerpy** [comment]
> The draft tokens are generated from a cached trie tree (each node is a token id). Currently the trie tree is constructed from prompts and responses on-the-fly, therefore it is friendly for deployment (neither additonal assist models nor head training), and works pretty well in real-life scenarios. A bit more, we also have probed Jaccobi iteraction to yield addtional drafts and will integrate into the repo soon(even though its speedup is marginal).

I'm also confused about how draft tokens were created.
Do you mean that this draft tokens is generated from previous prompts records? And not generated by model sampling is it?

**Comment by zheyishine** [comment]
Yes
